pip/pip-303.md

```diff
@@ -0,0 +1,207 @@
+
+# Motivation
+
+When a topic has a large number of producers or consumers (over 1k), querying the `pulsarAdmin.topics().getPartitionedStats()` interface is slow and the response size is also large.
+As a result, it's essential to give users the option of querying producer and consumer information.
+
+
+
+# Goals
+
+## In Scope
+
+Add the API for `org.apache.pulsar.client.admin.Topics`
+```java
+CompletableFuture<PartitionedTopicStats> getPartitionedStatsAsync(
+        String topic, boolean perPartition, GetStatsOptions options);
+
+CompletableFuture<TopicStats> getStatsAsync(String topic, GetStatsOptions options);
+```
+
+
+
+## Out of Scope
+
+None.
+
+
+# High Level Design
+
+Implement the `getPartitionedStatsAsync` method, and add the `exclusivePublishers` and `exclusiveConsumers` parameters to `{tenant}/{namespace}/{topic}/partitioned-stats` API in `PersistentTopics` and `NonPersistentTopics`.
+
+# Detailed Design
+
+## Design & Implementation Details
+
+
+Add two fields for `org.apache.pulsar.client.admin.GetStatsOptions`
+```java
+@Data
+@Builder
+public class GetStatsOptions {
+    /**
+     * Whether to exclusive publishers.
+     */
+    private final boolean exclusivePublishers;
+
+    /**
+     * Whether to exclusive subscriptions.
+     */
+    private final boolean exclusiveConsumers;
+    
+}
+```
+
+Implement the `getPartitionedStatsAsync` and `getStatsAsync` interface for `org.apache.pulsar.client.admin.internal.TopicsImpl`
+```java
+@Override
+public CompletableFuture<PartitionedTopicStats> getPartitionedStatsAsync(String topic,
+        boolean perPartition, boolean getPreciseBacklog, boolean subscriptionBacklogSize,boolean getEarliestTimeInBacklog,
+    boolean exclusivePublishers, boolean exclusiveConsumers){
+    TopicName tn=validateTopic(topic);
+    WebTarget path=topicPath(tn,"partitioned-stats");
+    path=path.queryParam("perPartition",perPartition)
+    .queryParam("getPreciseBacklog",getPreciseBacklog)
+    .queryParam("subscriptionBacklogSize",subscriptionBacklogSize)
+    .queryParam("getEarliestTimeInBacklog",getEarliestTimeInBacklog)
+    .queryParam("exclusivePublishers",exclusivePublishers)
+    .queryParam("exclusiveConsumers",exclusiveConsumers);
+}
+
+@Override
+public CompletableFuture<TopicStats> getStatsAsync(String topic, boolean getPreciseBacklog,
+        boolean subscriptionBacklogSize,
+        boolean getEarliestTimeInBacklog,
+        boolean exclusivePublishers, boolean exclusiveConsumers){
+        TopicName tn=validateTopic(topic);
+        WebTarget path=topicPath(tn,"stats")
+        .queryParam("getPreciseBacklog",getPreciseBacklog)
+        .queryParam("subscriptionBacklogSize",subscriptionBacklogSize)
+        .queryParam("getEarliestTimeInBacklog",getEarliestTimeInBacklog).
+        .queryParam("exclusivePublishers",exclusivePublishers)
+        .queryParam("exclusiveConsumers",exclusiveConsumers);
+}        
```

Comment:
Could you adjust the indent and add the blanks to make the code more readable?

Reply:
Thanks for your suggestion, done.